src/envoy/alts/tsi_transport_socket.cc

```diff
@@ -0,0 +1,295 @@
+      if (status != TSI_OK) {
+        ENVOY_CONN_LOG(
+            info, "TSI: unprotecting message failure {}, closing connection",
+            callbacks_->connection(), status);
+        result.action_ = Network::PostIoAction::Close;
+        return result;
+      }
+      buffer.add(unprotected_buffer, unprotected_buffer_size_to_send);
+      message_bytes += processed_message_size;
+      message_size -= processed_message_size;
+      ENVOY_CONN_LOG(debug, "TSI: unprotecting message processed: {}",
+                     callbacks_->connection(), processed_message_size);
+    }
+
+    result.bytes_processed_ = raw_read_buffer_.length() - message_size;
+    raw_read_buffer_.drain(result.bytes_processed_);
+
+    ASSERT(status == TSI_OK);
+  }
+
+  ENVOY_CONN_LOG(debug, "TSI: do read result action {} bytes {} end_stream {}",
+                 callbacks_->connection(), enumToInt(result.action_),
+                 result.bytes_processed_, result.end_stream_read_);
+  return result;
+}
+
+Network::IoResult TsiSocket::doWrite(Buffer::Instance &buffer,
+                                     bool end_stream) {
+  if (!handshake_complete_) {
+    Network::PostIoAction action = doHandshake();
+    if (action == Network::PostIoAction::Close) {
+      return {action, 0, false};
+    }
+  }
+
+  if (handshake_complete_) {
+    size_t message_size = buffer.length();
+    auto *message_bytes =
+        reinterpret_cast<unsigned char *>(buffer.linearize(message_size));
+
+    unsigned char protected_buffer[4096];
+    size_t protected_buffer_size = sizeof(protected_buffer);
+    tsi_result result = TSI_OK;
+
+    ENVOY_CONN_LOG(debug, "TSI: protecting message size: {}",
+                   callbacks_->connection(), message_size);
+
+    while (message_size > 0) {
+      size_t protected_buffer_size_to_send = protected_buffer_size;
+      size_t processed_message_size = message_size;
+      result = tsi_frame_protector_protect(
+          frame_protector_, message_bytes, &processed_message_size,
+          protected_buffer, &protected_buffer_size_to_send);
+      if (result != TSI_OK) {
+        ENVOY_CONN_LOG(
+            info, "TSI: protecting message failure {} wait until next write",
+            callbacks_->connection(), result);
+        break;
+      }
+      raw_write_buffer_.add(protected_buffer, protected_buffer_size_to_send);
+      message_bytes += processed_message_size;
+      message_size -= processed_message_size;
+
+      ENVOY_CONN_LOG(debug, "TSI: protecting message processed: {}",
+                     callbacks_->connection(), processed_message_size);
+    }
+
+    // Don't forget to flush.
+    if (message_size == 0) {
+      ENVOY_CONN_LOG(debug, "TSI: protecting message flush: {}",
+                     callbacks_->connection(), message_size);
+      size_t still_pending_size;
+      do {
+        size_t protected_buffer_size_to_send = protected_buffer_size;
+        result = tsi_frame_protector_protect_flush(
```

Comment:
this frame protector code really looks ugly.   Maybe you should wrap it with c++ object now

Reply:
Done